### PR TITLE
Implemented support for WorkflowSpec.ArtifactRepositoryRef

### PR DESCRIPTION
**Pull request by Ark-kun**
This change allows the workflow to specify the reference the configMap holding the artifact repository configuration.

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
spec:
  artifactRepositoryRef:
    configMap: artifact-repository
    key: config
...
```

Fixes https://github.com/argoproj/argo/issues/1332

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -164,6 +167,16 @@ func (woc *wfOperationCtx) operate() {
 
 	woc.setGlobalParameters()
 
+	if woc.wf.Spec.ArtifactRepositoryRef != nil {
+		repoReference := woc.wf.Spec.ArtifactRepositoryRef
+		repo, err := getArtifactRepositoryRef(woc.controller, repoReference.ConfigMap, repoReference.Key)
+		if err == nil {
+			woc.artifactRepository = repo
+		} else {
+			woc.log.Errorf("Failed to load artifact repository configMap: %+v", err)
```
Comment (jessesuen) [inline workflow/controller/operator.go]: I think we should fail the workflow if we can't dereference the repository.

**Comment by sarabala1979** [comment]
@Ark-kun  we almost ready to release 2.4 RC-1. Can you address Jesse comment and resolve the conflict?

**Comment by Ark-kun** [comment]
> @Ark-kun we almost ready to release 2.4 RC-1. Can you address Jesse comment and resolve the conflict?

I'll do this ASAP

File: workflow/controller/operator.go (hunk 1 above)
Comment (Ark-kun) [reply workflow/controller/operator.go]: Done.

**Comment by Ark-kun** [comment]
> @Ark-kun we almost ready to release 2.4 RC-1. Can you address Jesse comment and resolve the conflict?

Done.

**Pull request merged**